src/bits/fp2_array.rs

```diff
@@ -123,10 +125,111 @@ macro_rules! bit_array_impl {
+            // Any irreducible polynomial of order 8 will do, but this is the one we've chosen as a constant for our implementation of GF(2^8)
+            //
+            // We assume that: x^8 + x^4 + x^3 + x + 1 = 0
+            //
+            // Now there are roots of this polynomial, but no *real* roots, only complex roots. Let's call one of those roots α.
+            // Our Galois field is really represented by polynomials of α. So our input a was:
+            // α^7 + α^4 + α + 1
+            //
+            // Since we know that x^8 + x^4 + x^3 + x + 1 = 0, we can distribute out terms of this structure and replace them with zero.
+            //
+            // So the result of our multipliation was:
+            // 110100011111110
+            // which you can think of as:
+            // x^14 + x^13 + 0 + x^11 + 0 + 0 + 0 + x^7 + x^6 + x^5 + x^4 + x^3 + x^2 + x + 0
+            //
+            // We start by reducing the degree of the highest order element.
+            // We know that x^6 * (x^8 + x^4 + x^3 + x + 1) = 0, so we can subtract this value without changing anything
+            //
+            // The result is the same as
+            //      110100011111110
+            //  XOR 10001101
+            //
+            // Basically just bit-shift the irreducible polynomial up 6 bits and XOR it.
+            //
+            // We can continue this process until we have reached a polynomial where all terms are less than x^8.
+            //
+            // This defines a field, since all of the following properties hold:
+            //
+            // 1. Closure: For any two elements a and b in the field, a+b and ab are also in the field.
+            // 2. Associativity: Addition and multiplication are both associative, meaning that (a+b)+c = a+(b+c) and (ab)c = a(bc) for any elements a,b, and c in the field.
+            // 3. Commutativity: Addition and multiplication are both commutative, meaning that a+b = b+a and ab = ba for any elements a and b in the field.
+            // 4. Identity elements: There exist unique elements 0 and 1 in the field such that for any element a in the field, a+0 = a and a×1 = a.
+            // 5. Inverse elements: For any non-zero element a in the field, there exists a unique element -a such that a+(-a) = 0, and there exists a unique element a^(-1) such that a×a^(-1) = 1.
+            // 6. Distributivity: Multiplication distributes over addition, meaning that a(b+c) = ab+ac for any elements a, b, and c in the field.
+            //
+            // Tests of Associativity, Commutativity and Distributivity are below.
             impl std::ops::Mul for $name {
                 type Output = Self;
                 fn mul(self, rhs: Self) -> Self::Output {
-                    self & rhs
+                    debug_assert!(2 * $bits < u128::BITS);
+                    let a = <Self as Into<u128>>::into(self);
```

Comment:
Let's change this to
```suggestion
                    let a = <Self as Fp2Array>::as_u128(self);
```

Reply:
Why would you suggest this? Why not the more generic Into?